# Coffee Sales Analysis

![coffee_image](../assets/coffee-image.jpeg "alt-coffee_image")

## About Author

Author: Joshua Farara

Project: title

### Contact Info
Click on link below to contact/follow/correct me:

Email: joshua.farara@gmail.com

[LinkedIn](https://www.linkedin.com/in/joshuafarara/)

[Facebook](https://www.facebook.com/josh.farara/)

[Twitter](https://x.com/FararaTheArtist)

[Github](https://github.com/JoshuaFarara)


## About Data

Title: Dataset Title

Dataset: [Link](https://www.kaggle.com/datasets/ihelon/coffee-sales/data)

Description of data:

This dataset contains detailed records of coffee sales from a vending machine. It is intended for analysis of purchasing patterns, sales trends, and customer preferences related to coffee products. The dataset spans from March 2024 to June 2024, capturing daily transaction data. And it's continue added new information.

### Dataset Columns Names

Features:


### Metadata

Source:

Collection Methodology:

License:



### Task

Describe task at hand for this dataset.

### Objectives

Describe objective at had for this dataset.

### Kernel Version Used

Python==3.11.7

## Import Libraries

We will use the following libraries¶
1. Pandas: Data manipulation and analysis
2. Numpy: Numerical operations and calculations
3. Matplotlib: Data visualization and plotting
4. Seaborn: Enhanced data visualization and statistical graphics
5. Scipy: Scientific computing and advanced mathematical operations

In [49]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt
import seaborn as sns
import scipy as sp

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
import sys

# # Get the directory of the current script
# script_dir = os.path.dirname(os.path.abspath(__file__))

# # Change the working directory to the script's directory
# os.chdir(script_dir)

for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

## Data Loading and Exploration | Cleaning

### Load a CSV file then creating a dataframe

In [50]:
# Kaggle Notebook
# df = pd.read_csv('/kaggle/input/coffee-sales/index.csv')


#Local Machine Notebook
df = pd.read_csv('../../data/coffee_sales_data.csv')


### Set the option to show maximum columns

In [51]:
pd.set_option('display.max_columns', None) 
pd.set_option('display.max_rows', None)

### Get a sneak peek of data
The purpose of a sneak peek is to get a quick overview of the data and identify any potential problems or areas of interest

In [52]:
df.head(5)

date                 datetime cash_type                 card  money  \
0  2024-03-01  2024-03-01 10:15:50.520      card  ANON-0000-0000-0001   38.7   
1  2024-03-01  2024-03-01 12:19:22.539      card  ANON-0000-0000-0002   38.7   
2  2024-03-01  2024-03-01 12:20:18.089      card  ANON-0000-0000-0002   38.7   
3  2024-03-01  2024-03-01 13:46:33.006      card  ANON-0000-0000-0003   28.9   
4  2024-03-01  2024-03-01 13:48:14.626      card  ANON-0000-0000-0004   38.7   

     coffee_name  
0          Latte  
1  Hot Chocolate  
2  Hot Chocolate  
3      Americano  
4          Latte

### Let's see the column names

In [53]:
df.columns

Index(['date', 'datetime', 'cash_type', 'card', 'money', 'coffee_name'], dtype='object')

### Let's have a look on the shape of the dataset

In [54]:
print(f"The Number of Rows are {df.shape[0]}, and columns are {df.shape[1]}.")

The Number of Rows are 896, and columns are 6.


### Let's have a look on the columns and their data types using detailed info function

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 896 entries, 0 to 895
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         896 non-null    object 
 1   datetime     896 non-null    object 
 2   cash_type    896 non-null    object 
 3   card         807 non-null    object 
 4   money        896 non-null    float64
 5   coffee_name  896 non-null    object 
dtypes: float64(1), object(5)
memory usage: 42.1+ KB


### Count the missing values

In [56]:
df.isnull().sum()

date            0
datetime        0
cash_type       0
card           89
money           0
coffee_name     0
dtype: int64

## Cleaning Set 1

* There are 896 rows, and 6 columns in the dataset.

* The data type of all columns are objects except for df['money'] which is float.

* The columns in the datasets are:
    * 'date', 'datetime', 'cash_type', 'card', 'money', 'coffee_name'
    
* There are a few missing values in the dataset, which we will read in detail and deal with later on in the notebook.

* rename columns 'cash_type':'payment_type','card':'card_number', 'money':'amount_paid_uah'

* datetime can be split into two columns and drop column time, because data already has a column that satisfies date.

* money column after renaming 'amount_paid_uah' can derive an exchanged USD column for later comparison and analysis




### Task:

Clean the data by changing column names 


1. Change column names to appropriate names matching the data.
2. 

## Changing column names 

Changed: cash_type to payment_type since cash and card payments are accepted

In [57]:
df.rename(columns={'cash_type':'payment_type', 'card':'card_number', 'money':'amount_paid_uah'}, inplace=True)

In [58]:
df.columns

Index(['date', 'datetime', 'payment_type', 'card_number', 'amount_paid_uah',
       'coffee_name'],
      dtype='object')

## Creating Time Column

In [59]:
df[['new_date', 'time']] =df['datetime'].str.split(' ', n=1, expand=True)
df = df.drop(['new_date', 'datetime'], axis=1)
# df = ['date', 'time', 'payment_type', 'card_number', 'amount_paid_usd', 'coffee_name']

In [60]:
# check the data
df.head()


date payment_type          card_number  amount_paid_uah  \
0  2024-03-01         card  ANON-0000-0000-0001             38.7   
1  2024-03-01         card  ANON-0000-0000-0002             38.7   
2  2024-03-01         card  ANON-0000-0000-0002             38.7   
3  2024-03-01         card  ANON-0000-0000-0003             28.9   
4  2024-03-01         card  ANON-0000-0000-0004             38.7   

     coffee_name          time  
0          Latte  10:15:50.520  
1  Hot Chocolate  12:19:22.539  
2  Hot Chocolate  12:20:18.089  
3      Americano  13:46:33.006  
4          Latte  13:48:14.626

In [61]:
# df['time']

## Convert amount_paid_uah(formerly df['money']) to USD.

.024 USD  = 1 Ukrainian hryvnia

1 USD = 40.3317	Ukrainian hryvnia

In [62]:
UAH_to_USD = .024
df['amount_paid_usd'] = df['amount_paid_uah'] * UAH_to_USD

In [63]:
df.head(5)

date payment_type          card_number  amount_paid_uah  \
0  2024-03-01         card  ANON-0000-0000-0001             38.7   
1  2024-03-01         card  ANON-0000-0000-0002             38.7   
2  2024-03-01         card  ANON-0000-0000-0002             38.7   
3  2024-03-01         card  ANON-0000-0000-0003             28.9   
4  2024-03-01         card  ANON-0000-0000-0004             38.7   

     coffee_name          time  amount_paid_usd  
0          Latte  10:15:50.520           0.9288  
1  Hot Chocolate  12:19:22.539           0.9288  
2  Hot Chocolate  12:20:18.089           0.9288  
3      Americano  13:46:33.006           0.6936  
4          Latte  13:48:14.626           0.9288

## Analytical Questions

Product analysis:
1. Total Revenue

2. Total Sales per Coffee 

3. Average amount paid per cup 
 
4. Average amount paid per cup for each coffee type in UAH

5. Which coffee had the highest/lowest sales?


Dates and Time analysis:
3. Which coffee is sold in the morning, afternoon, evening?

4. Which coffee is sold most/least during the week/weekend?

5. Which dates have the highest sales?

6. What day's of the week do coffee sales occur the most?

7. What is the most frequent time of sales at this vending machine?

8. What are the sale amounts for the card numbers, frequency, total amount paid, coffee types?

9. Are there any holidays that cause a spike or decline in sales?

10. 

Consumer analysis:

11. What are the sale amounts for the card numbers, frequency, total amount paid, coffee types?

12. What is the ratio amongst card consumers and cash consumers?

13. Do cash consumers spend more then card counterparts?

Currency analysis:
14. What is the exchange rate between UAH and USD at the time of this data? Compared to the average?

15. 

### Coffee Sales Report

#### Product Analysis:

In [64]:
df['coffee_name'] = df['coffee_name'].astype('str') 

In [65]:
df[['coffee_name', 'amount_paid_uah']]

coffee_name  amount_paid_uah
0                  Latte            38.70
1          Hot Chocolate            38.70
2          Hot Chocolate            38.70
3              Americano            28.90
4                  Latte            38.70
5    Americano with Milk            33.80
6          Hot Chocolate            38.70
7    Americano with Milk            33.80
8                  Cocoa            38.70
9    Americano with Milk            33.80
10   Americano with Milk            33.80
11             Americano            28.90
12                 Latte            40.00
13   Americano with Milk            33.80
14   Americano with Milk            33.80
15             Americano            28.90
16   Americano with Milk            33.80
17             Americano            28.90
18                 Latte            40.00
19                 Latte            38.70
20               Cortado            28.90
21             Americano            28.90
22         Hot Chocolate            38.70
23                 Cocoa            38.70
24               Cortado            28.90
25   Americano with Milk            33.80
26   Americano with Milk            33.80
27         Hot Chocolate            38.70
28                 Latte            38.70
29                 Latte            38.70
30   Americano with Milk            33.80
31              Espresso            24.00
32                 Latte            38.70
33                 Latte            38.70
34                 Latte            38.70
35         Hot Chocolate            38.70
36                 Cocoa            38.70
37         Hot Chocolate            38.70
38         Hot Chocolate            38.70
39               Cortado            28.90
40         Hot Chocolate            38.70
41   Americano with Milk            35.00
42             Americano            28.90
43               Cortado            28.90
44            Cappuccino            38.70
45            Cappuccino            38.70
46                 Latte            40.00
47         Hot Chocolate            38.70
48                 Latte            38.70
49                 Latte            40.00
50             Americano            28.90
51   Americano with Milk            33.80
52               Cortado            28.90
53            Cappuccino            38.70
54             Americano            28.90
55             Americano            28.90
56             Americano            28.90
57   Americano with Milk            33.80
58                 Latte            38.70
59                 Latte            38.70
60         Hot Chocolate            40.00
61              Espresso            24.00
62             Americano            28.90
63            Cappuccino            38.70
64             Americano            28.90
65             Americano            28.90
66                 Latte            38.70
67   Americano with Milk            33.80
68               Cortado            28.90
69            Cappuccino            38.70
70   Americano with Milk            33.80
71         Hot Chocolate            38.70
72            Cappuccino            38.70
73         Hot Chocolate            38.70
74             Americano            30.00
75   Americano with Milk            35.00
76                 Latte            40.00
77              Espresso            24.00
78   Americano with Milk            35.00
79                 Latte            38.70
80               Cortado            28.90
81                 Latte            40.00
82            Cappuccino            40.00
83               Cortado            30.00
84             Americano            30.00
85            Cappuccino            38.70
86             Americano            28.90
87                 Latte            38.70
88               Cortado            28.90
89                 Latte            40.00
90             Americano            28.90
91               Cortado            28.90
92         Hot Chocolate            38.70
93   Americano with Milk            33.80
94             American

##### Plot total revenue

In [66]:
# 1. Total Revenue
total_revenue_uah = df['amount_paid_uah'].sum()
total_revenue_usd = df['amount_paid_usd'].sum()
print(f"Total Revenue (UAH): {total_revenue_uah}, Total Revenue (USD): {total_revenue_usd}")

Total Revenue (UAH): 30592.940000000002, Total Revenue (USD): 734.23056


In [67]:
# Plot total revenue
plt.figure(figsize=(5, 5))
plt.bar(['Total Revenue (UAH)', 'Total Revenue (USD)'], [total_revenue_uah, total_revenue_usd], color=['blue', 'green'])
plt.xlabel('Currency')
plt.ylabel('Total Revenue')
plt.title('Total Revenue in UAH and USD')
plt.savefig('./visualizations/Total Revenue in UAH and USD.jpg')
plt.close()

In [68]:
# Group data by coffee_name
grouped_coffee = df.groupby('coffee_name')

In [69]:
# 3. Total Sales per Coffee Type in UAH
total_amount_paid_uah = grouped_coffee['amount_paid_uah'].sum()
print(total_amount_paid_uah)

coffee_name
Americano              3786.42
Americano with Milk    6738.14
Cappuccino             6253.50
Cocoa                   995.66
Cortado                2422.80
Espresso                827.34
Hot Chocolate          2417.46
Latte                  7151.62
Name: amount_paid_uah, dtype: float64


In [70]:
# Plot total sales per coffee type in UAH
plt.figure(figsize=(5, 5))
plt.bar(total_amount_paid_uah.index, total_amount_paid_uah, alpha=0.7, color='skyblue')
plt.xlabel('Coffee Name')
plt.ylabel('Total Amount Paid (UAH)')
plt.title('Total Sales per Coffee Type in UAH')
plt.xticks(rotation=45)
plt.savefig('./visualizations/Total Sales per Coffee Type in UAH.jpg')
plt.close()

##### Total Revenue by Coffee Name Avanced Plotting of 

In [71]:
# Group by coffee_name and calculate total revenue
total_revenue_uah_coffee = df.groupby('coffee_name')['amount_paid_uah'].sum()
total_revenue_usd_coffee = df.groupby('coffee_name')['amount_paid_usd'].sum()

In [72]:
# Create a plot with dual y-axes
fig, ax1 = plt.subplots(figsize=(10, 5))

# Plot UAH data
color = 'tab:blue'
ax1.set_xlabel('Coffee Name')
ax1.set_ylabel('Total Amount Paid (UAH)', color=color)
bars1 = ax1.bar(total_revenue_uah_coffee.index, total_revenue_uah_coffee, alpha=0.7, label='UAH', color=color,  align='center')
ax1.tick_params(axis='y', labelcolor=color)

# Create a second y-axis for USD data
ax2 = ax1.twinx()
color = 'tab:green'
ax2.set_ylabel('Total Amount Paid (USD)', color=color)
bars2 = ax2.bar(total_revenue_usd_coffee.index, total_revenue_usd_coffee, alpha=0.7, label='USD', color=color, width=0.4)
ax2.tick_params(axis='y', labelcolor=color)

plt.xticks(rotation=45)

# Title and legend
fig.suptitle('Total Amount Paid for Each Coffee Name (UAH and USD)')
fig.legend([bars1, bars2], ['UAH', 'USD'], loc='upper left')

plt.savefig('./visualizations/Total Amount Paid for Each Coffee Name (UAH and USD).jpg')
# Show plot
plt.close()

##### Average amount paid per cup

In [73]:
# Average amount paid per cup
df['amount_paid_uah'].mean()

np.float64(34.14390625)

##### Average Amount Paid by Coffee Name UAH and USD

In [74]:
# 2. Average Amount Paid per Cup in UAH and USD
average_amount_paid_uah = grouped_coffee['amount_paid_uah'].mean()
average_amount_paid_usd = grouped_coffee['amount_paid_usd'].mean()
print(average_amount_paid_uah)
print(average_amount_paid_usd)

coffee_name
Americano              28.469323
Americano with Milk    33.192808
Cappuccino             38.131098
Cocoa                  38.294615
Cortado                28.503529
Espresso               23.638286
Hot Chocolate          38.372381
Latte                  38.243957
Name: amount_paid_uah, dtype: float64
coffee_name
Americano              0.683264
Americano with Milk    0.796627
Cappuccino             0.915146
Cocoa                  0.919071
Cortado                0.684085
Espresso               0.567319
Hot Chocolate          0.920937
Latte                  0.917855
Name: amount_paid_usd, dtype: float64


In [75]:
# Plot average amount paid per cup for each coffee type in UAH and USD
plt.figure(figsize=(5, 5))
plt.bar(average_amount_paid_uah.index, average_amount_paid_uah, alpha=0.7, label='UAH', color='blue')
plt.bar(average_amount_paid_usd.index, average_amount_paid_usd, alpha=0.7, label='USD', color='green', width=0.4)
plt.xlabel('Coffee Name')
plt.ylabel('Average Amount Paid')
plt.title('Average Amount Paid per Cup for Each Coffee Type (UAH and USD)')
plt.legend()
plt.xticks(rotation=45)
plt.savefig('./visualizations/Average Amount Paid per Cup for Each Coffee Type (UAH and USD).jpg')
plt.close()

##### Top Highest Revenue Earning Coffee Names

In [76]:
total_revenue_uah_coffee = df.groupby('coffee_name')['amount_paid_uah'].sum().reset_index()

In [77]:
total_revenue_sorted = total_revenue_uah_coffee.sort_values('amount_paid_uah', ascending=False)

In [78]:
# total_revenue_sorted = total_revenue_sorted.head()
print("Top Highest Revenue Earning Coffees (UAH) Table:")
total_revenue_sorted.head(8)

Top Highest Revenue Earning Coffees (UAH) Table:


coffee_name  amount_paid_uah
7                Latte          7151.62
1  Americano with Milk          6738.14
2           Cappuccino          6253.50
0            Americano          3786.42
4              Cortado          2422.80
6        Hot Chocolate          2417.46
3                Cocoa           995.66
5             Espresso           827.34

In [79]:
# Top Highest Revenue Earning Coffee Names
plt.figure(figsize=(5, 5))
plt.bar(total_revenue_sorted['coffee_name'], total_revenue_sorted['amount_paid_uah'])
plt.xlabel('Coffee Name')
plt.ylabel('Amount Paid (UAH)')
plt.xticks(rotation=90)
_ = plt.title('Top Highest Revenue Earning Coffee Names')
plt.savefig('./visualizations/Top Highest Revenue Earning Coffee Names.jpg')
plt.close()


In [80]:
# Print highest and lowest revenue earning coffee
highest_revenue_coffee = total_revenue_sorted.iloc[0]  # First row for highest
lowest_revenue_coffee = total_revenue_sorted.iloc[-1]   # Last row for lowest

print(f"Highest Revenue Earning Coffee (UAH): {highest_revenue_coffee['coffee_name']} - Amount: {highest_revenue_coffee['amount_paid_uah']}")
print(f"Lowest Revenue Earning Coffee (UAH): {lowest_revenue_coffee['coffee_name']} - Amount: {lowest_revenue_coffee['amount_paid_uah']}")


Highest Revenue Earning Coffee (UAH): Latte - Amount: 7151.62
Lowest Revenue Earning Coffee (UAH): Espresso - Amount: 827.34


##### Top 3 Coffees

In [81]:
top_3_revenue_coffees = total_revenue_sorted.head(3)
print("Top 3 Highest Revenue Earning Coffees (UAH):")
top_3_revenue_coffees

Top 3 Highest Revenue Earning Coffees (UAH):


coffee_name  amount_paid_uah
7                Latte          7151.62
1  Americano with Milk          6738.14
2           Cappuccino          6253.50

In [83]:
plt.figure(figsize=(5, 5))
plt.bar(top_3_revenue_coffees['coffee_name'], top_3_revenue_coffees['amount_paid_uah'], color='blue')
plt.xlabel('Coffee Name')
plt.ylabel('Total Amount Paid (UAH)')
plt.xticks(rotation=20)
plt.title('Top 3 Highest Revenue Earning Coffee Names')
plt.savefig('./visualizations/Top 3 Highest Revenue Earning Coffee Names.jpg')
plt.close()


### Dates and Time analysis: